### PR TITLE
[JENKINS-64911] Streamline validity check

### PR DESCRIPTION
**Pull request by bitwiseman**
# Description
See [JENKINS-64911](https://issues.jenkins-ci.org/browse/JENKINS-64911) for further information. 

Check validity was following its own special path, getting the password to create a hash, this resulted in a new call to `gitHub.checkApiUrlValidity()` each time the app password needed to be refreshed.  This resulted in users encountering the GitHub App issue we've seen with failure to authenticate much more often.

This change removes that special path and dependence on `getPassword()`. This still does address the underlying issue with App credentials, but so it goes.

# Submitter checklist
- [ ] Link to JIRA ticket in description, if appropriate.
- [ ] Change is code complete and matches issue description
- [ ] Automated tests have been added to exercise the changes
- [ ] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.

# Reviewer checklist
- [ ] Run the changes and verify that the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

# Documentation changes
- [ ] Link to jenkins.io PR, or an explanation for why no doc changes are needed

# Users/aliases to notify
@timja

### CODE REVIEW

**Review by timja — APPROVED** [review]

**Comment by bitwiseman** [comment]
NOTE: https://issues.jenkins.io/browse/JENKINS-62249 is the underlying source of this issue. If you want to collect logs as described there before trying this fix, that would be helpful.

**Comment by bitwiseman** [comment]
Improved tests, but not quite fixed. The failures are test issues not product issues.

**Comment by bitwiseman** [comment]
The more I think about it the more I think this will not solve the JENKINS-64911, though it should reduce it.  The underlying cause is likely the same as https://issues.jenkins.io/browse/JENKINS-62249.  I have a theory that there is a problem with clock skew.  I've pushed a couple of additional commits which might address clock skew.

**Comment by bitwiseman** [comment]
Figured it out, this is unrelated to JENKINS-62249.  

The [`GitHubAppCredentials.TokenProvider`](https://github.com/jenkinsci/github-branch-source-plugin/blob/63f92740f3903173ec9a79def4bfd868524674dc/src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubAppCredentials.java#L155) extends `GitHub.DependentAuthorizationProvider`.  That class has a specialized `GitHub` that ignores rate limit.  It does that for this specific reason, we do not want to sleep when using a JWT to request a `GHAppInstallationToken`.  The source of this issue is that `Connector.checkValidity` was not using the `GitHubAppCredentials.TokenProvider` code path, it was getting the password directly and ending up sleeping due to rate limit.  Not sure why that is happening, but this change removes that code path entirely and guarantees that the first call using any `GitHub` instance returned from `Connector` goes through the `TokenProvider`. 
However, it also highlights a flaw in the design - any call to `getPassword()` might encounter this problem. We need to make sure all calls to `getPassword()` use the `TokenProvider` code path.

**Comment by bitwiseman** [comment]
I've now made all code paths for token refresh use `GitHub.DependentAuthorizationProvider` derived instances.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java [hunk 1]
```diff
@@ -686,6 +668,21 @@ private static void removeAllUnused(long threshold) throws IOException {
                 }
             }
         }
+
+        public void verifyConnection() throws IOException {
+            synchronized (this) {
+                if (lastVerified > System.currentTimeMillis() - API_URL_REVALIDATE_MILLIS) {
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java:674]: well here be dragons.   the system clock can go backwards as well as forwards, and may not go forward by as much as you expect despite the amount of time passing.

TL;DR if revalidating is important to you based on some elapsed wall clock time only ever use `System.nanoTime`

Comment (bitwiseman) [reply src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java:674]: @jtnord 
Interesting. Found an article giving the trade-offs between the two: https://www.geeksforgeeks.org/java-system-nanotime-vs-system-currenttimemillis/ 

In this case, if verification occurs more or less often than expected in some edge cases that is okay.  We do verification in an attempt to give user meaningful errors when the server they are trying to reach is inaccessible or becomes inaccessible. If it isn't completely accurate that's okay. 

There are other areas where this information makes a difference.  Rate limit calculations for example. 

Thanks, I'll keep this in mind.

**Review by jtnord — APPROVED** [review]

**Review by car-roll — APPROVED** [review]
